src/java.base/share/classes/java/lang/reflect/code/op/CoreOps.java

```diff
@@ -1668,6 +1662,7 @@ public ArrayStoreOp transform(CopyContext cc, OpTransformer ot) {
 
             @Override
             public TypeDesc resultType() {
+                // resultType is void, but we still call resultType method because it has validations in it
```

Comment:
I think we should avoid specific validations in the call to `resultType()`, since it defers errors until the method is called.

In general i think we should separate out op specific validation into a separate method that ops can optionally implement (future work). The op constructors can perform some basic structural checks e.g., correct number of operands etc.

I have found checks on types when constructing ops (and by extension on a call to resultType) to sometimes be problematic, they can be fragile or too opinionated depending on how one consumes the model, and it seems they should reside in explicit validation functionality that can be optionally called.

Reply:
For now, should I validate during construction ?

Reply:
I suggest to remove checks from these checks, don't add them to the constructor, and add a @@@ note in the constructor. We need to revisit this area for construction/validation of all ops, where there are two entry points, from its definition and from its factory method (the latter is potentially more controlled than the former)